_posts/2022-09-28-run-ansible-playbooks.markdown

```diff
@@ -0,0 +1,104 @@
+_Project Flotta_ makes your life easier by supporting _Ansible_ playbook execution.
+How can we create an _Ansible_ playbook for the edge devices? How does the execution work in _Project Flotta_?
+his is what we will cover in this blog post.
+
+# Preliminary steps
+## 1- Define the Ansible Playbook
+First things first: we need to write your example _Ansible_ playbook.
+To keep things easy, let's say we want to create a txt file in some of our edge devices.
+
+```yaml
+---
+-  name: Hello Ansible Playbook
+   hosts: 127.0.0.1
+   gather_facts: false
+   
+   tasks:
+   - name: Create a file called '/tmp/hello.txt'
+     copy:
+       content: Hello from Project Flotta!
+       dest: /tmp/hello.txt
+```
+
+## 2 - Define the EdgeConfig
+Then, it's time to send it to the _Flotta Operator_ but... How?
+Easy! Let create an `EdgeConfig` CR! \
+(See [CRD Reference](https://project-flotta.io/documentation/v0_2_0/operations/crd.html#edgeconfig) for detailed description).
+
+```yaml
+apiVersion: management.project-flotta.io/v1alpha1
+kind: EdgeConfig
+metadata:
+  name: edgeconfig-sample
+spec:
+    edgePlaybook: 
+        playbooks: 
+            - content: LS0tCi0gIG5hbWU6IEhlbGxvIEFuc2libGUgUGxheWJvb2sKICAgaG9zdHM6IDEyNy4wLjAuMQogICBnYXRoZXJfZmFjdHM6IGZhbHNlCiAgIAogICB0YXNrczoKICAgLSBuYW1OiBDcmVhdGUgYSBmaWxlIGNhbGxlZCAnL3RtcC9oZWxsby50eHQnCiAgICAgY29weToKICAgICAgIGNvbnRlbnQ6IEhlbGxvIGZyb20gUHJvamVjdCBGbG90dGEhCiAgICAgICBkZXN0OiAvdG1wL2hlbGxvLnR4dAo=
+              timeoutSeconds: 10
+```
+
+The `content` item contains the `base64` encoding of our example playbook. It can be obtained using:
+
+```bash
+>> base64 << EOF
+---
+-  name: Hello Ansible Playbook
+   hosts: 127.0.0.1
+   gather_facts: false
+   
+   tasks:
+   - name: Create a file called '/tmp/hello.txt'
+     copy:
+       content: Hello from Project Flotta!
+       dest: /tmp/hello.txt
+EOF
+```
+## 3 - Label the edge devices
+To let the _Flotta Operator_ know that you want execute your `edgeconfig-sample` on one more specific edge devices, you must label them.
+
+```bash
+>> kubectl label edgedevice device1 config/device-by-config=edgeconfig-sample
+>> kubectl label edgedevice device3 config/device-by-config=edgeconfig-sample
+```
+In this case, we want to run the playbook on two edge devices: `device1` and `device3`
+## 4 - Apply the EdgeConfig to the cluster
+
+```bash
+>> kubectl apply -f edgeconfig-sample.yaml
+```
+
+# Architecture
+What happens when _Flotta Operator_ receives the `EdgeConfig`?
+
+It automatically creates a `PlaybookExecution` CR for each device that has been properly labelled.
+
+<img src="/assets/images/architecture_ansible_support.png" alt="High level architecture of Ansible support" width="1100"/>
+
+In this way it is possible to monitor the execution of the playbook on each edge device, indeed the edge device will update the status of the `PlaybookExecution` CR according with the Ansible playbook execution result.
```

Comment:
I would add what happens when a device is rebooted while playbook is being run and how many times plabook will be run on a device.

Reply:
Thanks for the suggestions!
> how many times plabook will be run on a device

Added a paragraph to describe the possible execution strategies

> when a device is rebooted while playbook is being run 

I'd like to wait for reviews in the device-worker PR so I can write a "matching" description of the rebooting case.
Then I'll ask you to re-review if you don't mind :)